### PR TITLE
[RFR] Convert GET_MANY accumulate saga to hooks

### LINKED ISSUE #3552: ReferenceArrayField causing infinite loop when empty collection
**What you were expecting:**
`ReferenceArrayField` loads associated records.

**What happened instead:**
`ReferenceArrayField` does initially load records, but gets caught in an infinite loop, calling `setState`.

**Steps to reproduce:**
Create `Show` view inside of a `TabbedShowLayout` (`SimpleShowLayout` also probably fine, but not tested) with a `ReferenceArrayField` containing an array of id's for associated records. When the view loads, it gets caught in infinite loop and will eventually lock up. This **ONLY** occurs when the collection is empty.

**Related code:**
Difficult to reproduce in sandbox, but can try to get something up if needed.

```
const UserShow = ({ record, ...props }) => (
  <Show
    {...props}
    title={<UserTitle />}
    aside={null}
    actions={<UserShowActions />}
  >
    <TabbedShowLayout>
      <Tab label="resources.users.tabs.user">
        <ReferenceArrayField
          label="Content Codes"
          reference="content/codes"
          source="content_code_ids"
          sort={{ field: 'code', order: 'ASC' }}
          allowEmpty
        >
          <SingleFieldList linkType="show">
            <ChipField source="code" />
          </SingleFieldList>
        </ReferenceArrayField>
      </Tab>
    </TabbedShowLayout>
  </Show>
)
```

**Environment**

* React-admin version: v3.0.0.alpha-2
* Last version that did not exhibit the issue (if applicable): v2.9.5
* React version: 16.9.0
* Browser: Chrome Version 76.0.3809.100
* Stack trace (in case of a JS error):
```
Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
    in ReferenceArrayField (at UserShow.js:57)
    in div (created by Labeled)
    in div (created by ForwardRef(FormControl))
    in ForwardRef(FormControl) (created by WithStyles(ForwardRef(FormControl)))
    in WithStyles(ForwardRef(FormControl)) (created by Labeled)
    in Labeled (created by WithStyles(Labeled))
    in WithStyles(Labeled) (created by Tab)
    in div (created by Tab)
    in span (created by Tab)
    in Tab (created by Context.Consumer)
    in translate(Tab) (at UserShow.js:44)
    in Route (created by TabbedShowLayout)
    in div (created by ForwardRef(CardContent))
    in ForwardRef(CardContent) (created by WithStyles(ForwardRef(CardContent)))
    in WithStyles(ForwardRef(CardContent)) (created by CardContentInner)
    in CardContentInner (created by WithStyles(CardContentInner))
    in WithStyles(CardContentInner) (created by TabbedShowLayout)
    in div (created by TabbedShowLayout)
    in TabbedShowLayout (created by Context.Consumer)
    in translate(TabbedShowLayout) (created by Context.Consumer)
    in withRouter(translate(TabbedShowLayout)) (at UserShow.js:43)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (created by ForwardRef(Card))
    in ForwardRef(Card) (created by WithStyles(ForwardRef(Card)))
    in WithStyles(ForwardRef(Card)) (created by ShowView)
    in div (created by ShowView)
    in div (created by ShowView)
    in ShowView (created by Show)
    in Show (at UserShow.js:37)
    in UserShow (created by WithPermissions)
    in WithPermissions (created by Context.Consumer)
    in Route (created by ResourceRoutes)
    in Switch (created by ResourceRoutes)
    in ResourceRoutes (created by Resource)
    in Resource (at App.js:69)
    in Route (created by RoutesWithLayout)
    in Switch (created by RoutesWithLayout)
    in RoutesWithLayout (created by Context.Consumer)
    in div (created by Layout)
    in main (created by Layout)
    in div (created by Layout)
    in div (created by Layout)
    in Layout (created by WithStyles(Layout))
    in WithStyles(Layout) (created by Context.Consumer)
    in withRouter(WithStyles(Layout)) (created by ConnectFunction)
    in ConnectFunction (created by LayoutWithTheme)
    in ThemeProvider (created by LayoutWithTheme)
    in LayoutWithTheme (at Layout.js:17)
    in Unknown (created by WithStyles(Component))
    in WithStyles(Component) (created by Context.Consumer)
    in Route (created by CoreAdminRouter)
    in Switch (created by CoreAdminRouter)
    in div (created by CoreAdminRouter)
    in CoreAdminRouter (created by ConnectFunction)
    in ConnectFunction
    in ConnectFunction (created by Context.Consumer)
    in Route (created by CoreAdmin)
    in Switch (created by CoreAdmin)
    in Router (created by ConnectedRouter)
    in ConnectedRouter (created by Context.Consumer)
    in ConnectedRouterWithContext (created by ConnectFunction)
    in ConnectFunction (created by CoreAdmin)
    in TranslationProviderView (created by ConnectFunction)
    in ConnectFunction (created by CoreAdmin)
    in Provider (created by CoreAdmin)
    in CoreAdmin (at App.js:39)
    in App (at src/index.js:7)```

### PR DESCRIPTION
**Pull request by fzaninotto**
Using a hook in `ReferenceField` allows to handle corner cases that were impossible to deal with using sagas. Most notably, `ReferenceField` now works properly when the `dataProvider` fails or returns nothing for the `GET_MANY` call.

This is mostly a refactoring with few visible changes for the end user, labeled 'NEW' below.

- [x] Create `useGetMany` hook based on `crudGetManyAccumulate` action logic
- [x] Add unit test for `useGetMany`
- [x] Use the new hook in `useReference`
- [x] NEW Hide loader in `<ReferenceField>` if the `dataProvider` call fails or returns nothing
- [x] NEW Display an error icon in `<ReferenceField>` if the `dataProvider` fails
- [x] Use the new hook in `useReferenceArrayFieldController`
- [ ] ~~Use the new hook in `ReferenceArrayInputController`~~ (too long, to be done in another PR)
- [x] Fix tests 

Closes #3552, #3451, #2388, #854.

To be done in another PR:

- Create `useGetMatching` hook based on `crudGetMatchingAccumulate` action logic
- Add unit test for `useGetMatching`
- Use the new hook in `ReferenceInput`, `ReferenceArrayInput`, `SelectArrayInput` and `SelectInput`
- Deprecate `accumulate` saga

This change is backwards compatible with the `accumulate` saga, i.e. dispatching `crudGetManyAccumulate` actions still works with the saga. The new `useGetMany` hook does not dispatch such actions.

### CODE REVIEW

File: packages/ra-core/src/controller/useReference.ts [hunk 1]
```diff
@@ -39,44 +34,17 @@ export interface UseReferenceProps {
...
-            return referenceState && referenceState.data[id];
-        },
-        [id, reference]
-    );
-
-    return useSelector(getReferenceRecord);
+export const useReference = ({ reference, id }: Option): UseReferenceProps => {
+    const { data, error, loading, loaded } = useGetMany(reference, [id]);
+    if (error) {
+        console.log(error); // FIXME what should we do?
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useReference.ts]: Fixme

File: packages/ra-ui-materialui/src/field/ReferenceField.js [hunk 2]
```diff
@@ -138,9 +138,17 @@ export const ReferenceFieldView = ({
     ...rest
 }) => {
     const classes = useStyles({ classes: classesOverride });
-    if (loading) {
+    if (!loaded) {
         return <LinearProgress />;
     }
+    if (error) {
+        return (
+            <Error aria-errormessage="Error" color="error" fontSize="small" />
```
Comment (djhi) [inline packages/ra-ui-materialui/src/field/ReferenceField.js]: Shouldn't the message be the actual error (translated even)?

**Comment by fzaninotto** [comment]
Switching to RFR

**Comment by fzaninotto** [comment]
oops, sorry, I missed a side effect, back to WIP

**Comment by fzaninotto** [comment]
Well, that's in `ReferenceArrayInputController`, so not in this PR. Back to RFR.

File: packages/ra-ui-materialui/src/field/ReferenceField.js [hunk 3]
```diff
@@ -144,7 +144,11 @@ export const ReferenceFieldView = ({
     }
     if (error) {
         return (
-            <Error aria-errormessage="Error" color="error" fontSize="small" />
+            <Error
+                aria-errormessage={error.message ? error.message : error}
```
Comment (djhi) [inline packages/ra-ui-materialui/src/field/ReferenceField.js]: Shouldn't we translate this error ?

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceField.js]: Ideally, yes. In a future PR? I'd like to have this pr merged quickly to release a new alpha soon.

**Pull request merged**